File: LongestSong.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.rdd.RDD
import scala.util.Try
import java.io.StringReader
import org.apache.commons.csv.{CSVFormat, CSVParser}
import CSVUtil._

object LongestSong {
  def main(args: Array[String]): Unit = {
    val start: Long = System.nanoTime()
    //  Initialize Spark
    val conf = new SparkConf()
      .setAppName("LongestSongFinder")
      .setMaster("local[*]") // Use all CPU cores
    val sc = new SparkContext(conf)

    val inputFile = "src/main/resources/ds2Scala.csv"
    val linesRDD: RDD[String] = sc.textFile(inputFile)

    //  Merge multi-line records
    val mergedRDD: RDD[String] = linesRDD.coalesce(1).mapPartitions(mergeMultilineCSV)

    //  Function to parse CSV lines safely
    def parseCSVLine(line: String): Option[(String, Int)] = {
      Try {
        val reader = new StringReader(line)
        val csvParser = new CSVParser(reader, CSVFormat.DEFAULT.withQuote('"'))
        val row = csvParser.iterator().next()
        reader.close()
        csvParser.close()

        if (row.size() >= 7) {
          val title = row.get(0).trim
          val lyrics = row.get(6).trim
          val numLines = lyrics.split("\\R").length // Handles \n, \r\n, etc.
//          val numLines = lyrics.split(" ").length
          Some(title, numLines)
        } else None
      }.getOrElse(None)
    }

    //  Parse the merged RDD
    val parsedRDD: RDD[(String, Int)] = mergedRDD.flatMap(parseCSVLine)

    //  Find the longest song
    val longestSong: (String, Int) = parsedRDD.reduce((a, b) => if (a._2 > b._2) a else b)

    val end: Long = System.nanoTime()
    val duration: Double = (end - start) / 1e9 // Convert to seconds

    //  Print the result
    println("\nLongest Song in Dataset ")
    println(s"${longestSong._1}: ${longestSong._2} lines")
    println(s"\nTotal Execution Time: $duration seconds")

    //  Stop Spark
    sc.stop()

//    Longest Song in Dataset by line breaks \\R
//      Patient Protection and Affordable Care Act ObamaCare: 29251 lines
//      Total Execution Time: 142.511299042 seconds
//
//    Longest Song in Dataset by spaces
//      Patient Protection and Affordable Care Act ObamaCare: 365904 spaces
//      Total Execution Time: 141.570158917 seconds

  }
}

File: LongestSong.py
import csv
import sys
import csv

line_no=0
csv.field_size_limit(sys.maxsize)
max_lines=0
max_song=""
with open('ds29GB.csv', newline='') as csvfile:
    filereader = csv.reader(csvfile, delimiter=",", quotechar='"')
    for row in filereader:
        line_no+=1
        # if line_no%10000==0:
        #     print(row)
        lyrics=row[6]
        title=row[0]
        lines = lyrics.split(' ')#Using spaces can use \n
        no_of_lines=len(lines)
        if no_of_lines > max_lines:
            max_lines=no_of_lines
            max_song=title
        # if line_no > 50000:
        #       break
print(line_no)
print (max_lines)
print(max_song)
       
# Simple example of finding the song with the most lines in it using Highlevel language (Python)
#     Result using \n as split:
#         Current longest: Patient Protection and Affordable Care Act ObamaCare
#         Time taken: 4min 18.8sec
#         Line Count:29251
#         Total rows searched: 5913412
#     Result using " " as split:
#         Current longest: Patient Protection and Affordable Care Act ObamaCare
#         Time taken: 5min 33.3sec
#         Space Count:365904
#         Total rows searched: 5913412

# Taking this as the longest as it simplity goes through them all

# SQL Query: SELECT title, 
#                 LENGTH(lyrics) - LENGTH(REPLACE(lyrics, ' ', '')) + 1 AS word_count
#             FROM songs
#             ORDER BY word_count DESC
#             LIMIT 1;
